### PR TITLE
Refactor text splitters split and merge

### PR DESCRIPTION
**Pull request by IlyaMichlin**
#9843

Replace this entire comment with:
  - Description: a description of the change, 
  - Issue: the issue # it fixes (if applicable),
  - Dependencies: any dependencies required for this change,
  - Tag maintainer: for a quicker response, tag the relevant maintainer (see below),
  - Twitter handle: we announce bigger features on Twitter. If your PR gets announced and you'd like a mention, we'll gladly shout you out!

Please make sure your PR is passing linting and testing before submitting. Run `make format`, `make lint` and `make test` to check this locally.

See contribution guidelines for more information on how to write/run tests, lint, etc: 
https://github.com/hwchase17/langchain/blob/master/.github/CONTRIBUTING.md

If you're adding a new integration, please include:
  1. a test for the integration, preferably unit tests that do not rely on network access,
  2. an example notebook showing its use. These live is docs/extras directory.

If no one reviews your PR within a few days, please @-mention one of @baskaryan, @eyurtsev, @hwchase17, @rlancemartin.
 -->

### CODE REVIEW

File: libs/langchain/langchain/text_splitter.py [hunk 1]
```diff
@@ -284,20 +266,29 @@ class CharacterTextSplitter(TextSplitter):
...
         """Split incoming text and return chunks."""
-        # First we naively split the large input into a bunch of smaller ones.
-        separator = (
-            self._separator if self._is_separator_regex else re.escape(self._separator)
-        )
-        splits = _split_text_with_regex(text, separator, self._keep_separator)
-        _separator = "" if self._keep_separator else self._separator
-        return self._merge_splits(splits, _separator)
+        splits = _split_text_with_regex(text, self._separator)
+        return self._merge_splits(splits)
```
Comment (leo-gan) [inline libs/langchain/langchain/text_splitter.py:291]: Those are pretty heavy refactoring.
Is it possible to support them with unit tests to make sure everything works well?

Comment (baskaryan) [reply libs/langchain/langchain/text_splitter.py:291]: agreed, some unit tests would be great!

Comment (IlyaMichlin) [reply libs/langchain/langchain/text_splitter.py:305]: Definetely yes!
This is still in development

**Comment by leo-gan** [comment]
@baskaryan Please, review it. Thanks!

**Comment by leo-gan** [comment]
@baskaryan Please, review it. Thanks!

**Comment by leo-gan** [comment]
@baskaryan Please, review it. Thanks!

**Comment by IlyaMichlin** [comment]
@baskaryan this PR fixes the issue with adding the separators to the splitted text. If you look on the tests, you will see that there are many splits happening mid-words. I fix this issue as well.
This is done by moving all the splits to use regex instead of raw strings. This opens up the posibility to define structures on which to split programming language text.
Will be glad if this PR can be reviewed.

File: libs/langchain/tests/unit_tests/test_text_splitter.py [hunk 2]
```diff
@@ -240,32 +235,31 @@ def __test_iterative_text_splitter(chunk_size: int, keep_separator: bool) -> Lis
...
+    splitter = RecursiveCharacterTextSplitter(chunk_size=chunk_size, chunk_overlap=1)
     output = splitter.split_text(text)
     expected_output = [
-        "Hi.",
-        "I'm",
+        "Hi.\n\nI'm H",
         "Harrison.",
-        "How? Are?",
-        "You?",
+        ".\n\nHow? Ar",
```
Comment (baskaryan) [inline libs/langchain/tests/unit_tests/test_text_splitter.py:247]: hm seems like the original behavior was nicer?

Comment (IlyaMichlin) [reply libs/langchain/tests/unit_tests/test_text_splitter.py]: Good point. Fixed that.

**Comment by hwchase17** [comment]
closing due to large merge conflicts

**Pull request closed without merging**